aiida/backends/managers/profile_access.py

```diff
@@ -0,0 +1,222 @@
+            pass
+
+    If the profile is already locked or is currently in use, an exception is raised.
+
+    Access and locks of the profile will be recorded in a directory with files with a ``.pid`` and ``.lock`` extension,
+    respectively. In principle then, at any one time, there can either be a number of pid files, or just a single lock
+    file. If there is a mixture or there are more than one lock files, we are in an inconsistent state.
+    """
+
+    def __init__(self, profile: Profile):
+        """Class that manages access and locks to the given profile.
+
+        :param profile: the profile whose access to manage.
+        """
+        from aiida.manage.configuration.settings import ACCESS_CONTROL_DIR
+
+        type_check(profile, Profile)
+        self.profile = profile
+        self.process = psutil.Process(os.getpid())
+        self._dirpath_records = ACCESS_CONTROL_DIR / profile.name
+        self._dirpath_records.mkdir(exist_ok=True)
+
+    def request_access(self) -> None:
+        """Request access to the profile.
+
+        If the profile is locked, a ``LockedProfileError`` is raised. Otherwise a PID file is created for this process
+        and the function returns ``None``. The PID file contains the command of the process.
+
+        :raises ~aiida.common.exceptions.LockedProfileError: if the profile is locked.
+        """
+        error_message = (
+            f'process {self.process.pid} cannot access profile `{self.profile.name}`'
+            f'because it is being locked.'
+        )
+        self._raise_if_locked(error_message)
+
+        filepath_pid = self._dirpath_records / f'{self.process.pid}.pid'
+        filepath_tmp = self._dirpath_records / f'{self.process.pid}.tmp'
+
+        try:
+            # Write the content to a temporary file and then move it into place with an atomic operation.
+            # This prevents the situation where another process requests a lock while this file is being
+            # written: if that was to happen, when the locking process is checking for outdated records
+            # it will read the incomplete command, won't be able to correctly compare it with its running
+            # process, and will conclude the record is old and clean it up.
+            filepath_tmp.write_text(str(self.process.cmdline()))
+            os.rename(filepath_tmp, filepath_pid)
+
+            # Check again in case a lock was created in the time between the first check and creating the
+            # access record file.
+            error_message = (
+                f'profile `{self.profile.name}` was locked while process '
+                f'{self.process.pid} was requesting access.'
+            )
+            self._raise_if_locked(error_message)
+
+        except Exception as exc:
+            filepath_tmp.unlink(missing_ok=True)
+            filepath_pid.unlink(missing_ok=True)
+            raise exc
+
+    @contextlib.contextmanager
+    def lock(self):
```

Comment:
The thing is that at this point I think we would be introducing a new library not only at normal cost (i.e. dependency maintenance), but also with the extra of having to coordinate it awkwardly for our use case (which adds another point of possible break), and with little to nothing to gain:

- Our `lock` method still has to check the access files which is basically 80% of its logic, so not a lot of simplification there. The internal call to their `lock` would only be replacing the file creation basically (which should not have a lot of platform dependency issues).
- The typical benefit of "outsourcing" the effort of having to deal with maintaining the implementation is reduced given that we anyways would need to maintain the other half of tracking "non-exclusive" access.
- With respect to the difference in behavior (our `lock` currently raises if you try to lock a profile that isalready locked, their locking mechanism has the process waiting for the other lock to release) I think I actually prefer ours. The situation where this is relevant is when a racing condition triggers and we have two processes trying to acquire the lock at almost the same time. I think I personally prefer our resolution to this situation.

I don't know, maybe I'm underestimating the possible complications of locking with files, but I don't feel this is a good tradeoff. @sphuber you are quite familiar with the code by now, do you feel like we could gain something from trying to fit in the `py-filelock` in this design?

Reply:
I see both points. Normally, for such an important thing, I would also be tempted to use the external package, but I think you expressed the counter-arguments pretty well. We could indeed use it for our own `lock` implementation, but for the `request_access` we would still have our own implementation.

There would still be a robustness advantage in replacing our own locking implementation with the `filelock` package in the `lock` method. If we simply use their `FileLock`, we don't have to worry about racing conditions and robustness. As simple as it may seem, it is often quite tricky to be sure there are no subtle issues. The problem is that we cannot use the process ID in the filename, because the lock filename should be identical for all processes. But we can _also_ not write the process ID _in_ the lockfile, which the docs of `filelock` clearly state not to do. If we want to keep the functionality to be able to tell which process already has a lock when another tries to acquire it, we would have to start writing that to a separate file. But that has problems in and of its own.

TLDR: I think it would be preferable to use the package, but I think this is not possible due to our requirements of having non-exclusive locks and wanting to notify a user which processes are accessing it or locking it if a lock is denied. Only if we are willing to give up this user functionality, can we simplify the implementation and use the package. So in the end, this is the decision we have to take I believe.

Reply:
I would suggest move forward with the current implementation, ensure that is extremely well tested (that's one of those things that should have 100% coverage IMO), and then maybe we can spend 1-2hrs trying to refactor it with the libraries that I have mentioned. If it is too complicated, then we just stick with the current approach until we run into problems.